### PR TITLE
Add `flake.nix`

### LINKED ISSUE #27: Add a `flake.nix`
Convert the `default.nix` to a `flake.nix` such that `nix build` builds the package and `nix develop` puts us in a development shell where both `cabal build` and `haskell-language-server` commands run and succeed.

Ref:

- https://nixos.wiki/wiki/Flakes
- https://nixos.wiki/wiki/Haskell

### PR DESCRIPTION
**Pull request by conscious-puppet**
Closes #27

### CODE REVIEW

File: hie.yaml [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+              - "-package Cabal"
+              - "-package base"
+    - path: ./
+      config:
+        cradle:
+          cabal:
+            - path: "./src"
+              component: "lib:jrec"
+            - path: "./test"
+              component: "test:jrec-test"
```
Comment (srid) [inline hie.yaml:17]: You can simplify by this by getting rid of `Setup.hs` (not really required?) and using:

```yaml
cradle:
  cabal:
```

https://github.com/srid/ema/blob/master/hie.yaml

Comment (srid) [reply hie.yaml:17]: You also need `other-modules` https://gist.github.com/srid/27237dfd02e64150e5fa563c2ffff590

Comment (abshekh) [reply hie.yaml]: i have removed `Setup.hs`. It is used for building and packaging, but we are using nix. So you are right, it is not required.

**Review by srid — APPROVED** [review]
LGTM

I'm just waiting to confirm if we still care about supporting older GHCs.

**Comment by srid** [comment]
I'll merge this. Aeson 1.x + GHC 8.8 support is being added in a different PR: https://github.com/juspay/jrec/pull/29

**Pull request merged**